File: Python/VM.py
import numpy as np
import sys
import time
from math import log


class CacheLine:
    #simule une ligne de Cache

    def __init__(self, taille):
        self.use = 0
        self.valid = 0
        self.tag = 0
        self.data = [0] * taille

class Cache:
    #simulation d'un Cache à correspondance directe

    def __init__(self, S, B, memorySize):

        #initialisation quantités S,E,B,m avec E=1 et m= log2(taille.memoire)
        self.nbSets = S
        self.setSize = int(log(S,2)) #taille du champ set
        self.nbLines = 1
        self.blockSize = B
        self.nbBits = int(log(memorySize,2))

        self.mask_set = 2**(self.setSize+1)-1

        self.lines = [CacheLine(self.blockSize) for i in range(memorySize // self.blockSize)] #création des lignes de notre Cache

    def getTag(self, adresse):
        #retourne le tag d'une adresse
        return adresse >> (self.blockSize+self.setSize)

    def getSet(self, adresse):
        numSet =  (adresse >> self.blockSize) & self.mask_set
        return self.lines[numSet]

    def read(self, addresse):
        #Lit un bloc de mémoire
        tagA = getTag(addresse)  # Tag of cache line
        setA = getSet(addresse)  # Set of cache lines

        if (self.lines[setA].tag == tagA and self.lines[setA].valid):
            return self.lines[setA].data

    def write(self, adresse, donnee):
        #ecrit une donnee dans un bloc memoire
        tagA = getTag(addresse)  # Tag of cache line
        setA = getSet(addresse)  # Set of cache lines

        if (self.lines[setA].tag == tagA and self.lines[setA].valid):
            self.lines[setA].data == donnee





class VM:

    def __init__(self):

        self.INSTR_LIM        = 100
        self.NUM_REGS         = 32
        self.TAILLE_MAX       = 100
        self.MAX_INSTRUCTIONS = 100

        self.regs    = np.zeros((self.NUM_REGS), dtype=int)
        self.data    = np.zeros((1,128))
        self.program = []
        self.pc      = -1

        self.running = 1

        self.instrNum = 0
        self.reg1     = 0
        self.reg2     = 0
        self.imm      = 0
        self.o        = 0
        self.a        = 0
        self.n        = 0

        self.cache = Cache(16,8,1024)

    def fetch(self):
        self.pc += 1
        return self.program[self.pc]

    def decode(self):
        #decodage d'une ligne ecrite au format hexadecimal
        self.instrNum = (self.instr & (31 << 27)) >> 27

        if (self.instrNum == 15):
            self.imm      = (self.instr & (1<<26)  )  >> 26
            self.o        = (self.instr & 0x3FFFE0 )  >>  5
            self.reg1     = (self.instr & 0x00001F)
        elif (self.instrNum == 16):
            self.reg1     = (self.instr & 0x7C00000)  >> 22
            self.a        = (self.instr & 0x3FFFFF )
        elif (self.instrNum == 17):
            self.reg1     = (self.instr & 0x7C00000)  >> 22
            self.a        = (self.instr & 0x3FFFFF )
        elif (self.instrNum == 18):
            self.reg1     = (self.instr & 0x7C00000)  >> 22
            self.imm        = (self.instr & 0x3FFFFFF)
        else:
            self.reg1     = (self.instr & 0x7C00000)  >> 22
            self.imm      = (self.instr & 0x200000 )  >> 21
            self.o        = (self.instr & 0x1FFFE0 )  >>  5
            self.reg2     = (self.instr & 0X1F     )

    def evaluate(self, renouvelable = 1):
        #petite precision : le parametre "renouvelable" empeche l'appel recursif infini de la fonction
        print("evaluation ===> instruNum = {}".format(self.instrNum))
        if (self.instrNum == 0):
            # halt
            print("stop")
            self.running = 0
        elif self.instrNum == 1:
            # add
            if (self.imm == 1):
                print( "add r{} #{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] + self.o
            else:
                print( "add r{} r{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] + self.regs[self.o]
        elif self.instrNum ==  2:
            # add
            if (self.imm == 1):
                print( "sub r{} #{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] - self.o
            else:
                print( "sub r{} r{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] - self.regs[ self.o ]
        elif self.instrNum == 3:
            # mult
            if (self.imm == 1):
                print( "mult r{} #{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] * self.o
            else:
                print( "mult r{} r{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] * self.regs[ self.o ]
        elif self.instrNum == 4:
            # div
            if (self.imm == 1):
                print( "div r{} #{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] / self.o
            else :
                print( "div r{} r{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] / self.regs[ self.o ]
        elif self.instrNum == 5:
            # and
            if (self.imm == 1):
                print( "and r{} #{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] & self.o
            else :
                print( "and r{} r{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] & self.regs[ self.o ]

        elif self.instrNum == 6:
            #or
            if (self.imm == 1) :
                print( "or r{} #{} r{}".format( self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] | self.o
            else :
                print( "or r{} r{} r{}".format( self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] | self.regs[ self.o ]
        elif self.instrNum == 7:
            #xor
            if (self.imm == 1):
                print( "xor r{} #{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] ^ self.o

            else :
                print( "xor r{} r{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] ^ self.o
        elif self.instrNum == 8:
            # shl
            if (self.imm == 1):
                print( "shl r{} #{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] << self.o;
            else:
                print( "shl r{} r{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] << self.regs[ self.o ]

        elif self.instrNum == 9:
            # shr
            if (self.imm == 1):
                print( "shr r{} #{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] >> self.o

            else :
                print( "shr r{} r{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] >> self.regs[ self.o ]

        elif self.instrNum == 10:
            # slt
            if (self.imm == 1):
                print( "slt r{} #{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] < self.o

            else:
                print( "slt r{} r{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] < self.regs[ self.o ]

        elif self.instrNum == 11:
            # sle
            if (self.imm == 1):
                print( "sle r{} #{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] <= self.o

            else:
                print( "sle r{} r{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] <= self.regs[ self.o ]

        elif self.instrNum == 12:
            #seq
            if (self.imm == 1):
                print( "seq r{} #{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] == self.o

            else:
                print( "seq r{} r{} r{}".format(self.reg1, self.o, self.reg2) )
                self.regs[ self.reg2 ] = self.regs[ self.reg1 ] == self.regs[ self.o ]

        elif self.instrNum == 13:
            #load
            print( "load r{} #{} r{}".format(self.reg1, self.o, self.reg2))
            self.regs[ self.reg2 ] = self.regs[self.reg1 + self.o]

            self.cache.read(self.reg1 + self.o)


        elif self.instrNum == 14:
            #store
            print( "store r{} #{} r{}".format(self.reg1, self.o, self.reg2))
            self.regs[ self.reg1 + self.o ] = self.regs[self.reg2]

            self.cache.write(self.reg1 + self.o , self.regs[self.reg2])


        elif self.instrNum == 15:
            # jmp
            print( "jmp r{} L{}".format(self.reg1, self.o ))
            self.regs[self.reg1] = self.pc + 1
            self.pc = self.o
            self.instr = self.program[self.o]
            self.decode()
            if(renouvelable==1):
                self.evaluate(0)

        elif self.instrNum == 16:
            # braz
            print("braz r{} L{}".format(self.reg1, self.a))
            if (self.regs[self.reg1] == 0):
                self.instr = self.program[self.a]
                self.pc = self.a
                self.decode()
                if(renouvelable==1):
                    self.evaluate(0)

        elif self.instrNum == 17:
            #branz
            print("branz r{} L{}".format(self.reg1, self.a))
            if (self.regs[self.reg1] != 0):
                self.instr = self.program[self.a]
                self.pc = self.a
                self.decode()
                if(renouvelable==1):
                    self.evaluate(0)

        elif self.instrNum == 18:
            #scall

            print("scall n {}".format(self.imm))

            if (self.imm == 0):
                self.data = input("Donnez un nombre : ")
                self.regs[self.reg1] = int(self.data)
            else:
                print("Dans le registre {} : {}".format(self.reg1, self.regs[self.reg1]))



    """ display all registers as 4-digit hexadecimal words """
    def showRegs(self):
      print( "self.regs = " )

      for i in range(self.NUM_REGS // 4):
        print( "{:08x} {:08x} {:08x} {:08x} ".format(int(self.regs[ 4*i ]), int(self.regs[4*i+1]), int(self.regs[4*i+2]), int(self.regs[4*i+3])))

    def run(self):
        while( self.running ):
            self.showRegs()
            self.instr = self.fetch()
            self.decode()
            self.evaluate()
        self.showRegs()


if __name__=='__main__':

    tps1 = time.clock()

    VirtualMachine = VM()

    fichier = open(sys.argv[1], 'r')
    if (fichier == None):
        print("Erreur de lecture de fichier")
    else:
        fichier_chaine = fichier.readlines()

    for ind,elem in enumerate(fichier_chaine):
        VirtualMachine.program.append(int("0x"+elem[:-1],0))
    print(VirtualMachine.program)

    VirtualMachine.run()

    tps2 = time.clock()

    print("temps d'execution du program : {} sec".format(tps2-tps1))

File: C/VM.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#define INSTR_LIM  100
#define NUM_REGS   32
#define TAILLE_MAX 100
#define MAX_INSTRUCTIONS 100

unsigned regs[ NUM_REGS ];
char data[128];
unsigned long int program[MAX_INSTRUCTIONS];
/* program counter */
int pc = 0;

/* fetch the next word from the program */
unsigned long int fetch()
{

  return program[ pc++ ];
}

/* instruction fields */
unsigned int instrNum = 0;
int reg1     = 0;
int reg2     = 0;
int imm      = 0;
int o        = 0;
int a        = 0;
int n        = 0;

/* decode a word */
void decode( unsigned long int instr )
{

  instrNum = 0;
  reg1     = 0;
  reg2     = 0;
  imm      = 0;
  o        = 0;
  a        = 0;
  n        = 0;

  printf("instr        : %lu\n", instr);
  printf("instr & mask : %lu\n", instr & (31<<27));

  instrNum = (instr & (31<<27)) >> 27;
  //printf("numero d'instruction : %x\n instr & mask : %x \n instr&mask>> : %x \n", instrNum, instr & 0xFFFFFFFF, (instr & 0xFFFFFFFF) >> 27);
  switch (instrNum) {
    case 15:
      imm      = (instr & (1<<26)  )  >> 26;
      o        = (instr & 0x3FFFE0 )  >>  5;
      reg1     = (instr & 0x00001F)       ;
      break;
    case 16:
      reg1     = (instr & 0x7C00000)  >> 26;
      a        = (instr & 0x3FFFFF )       ;
      break;
    case 17:
      reg1     = (instr & 0x7C00000)  >> 26;
      a        = (instr & 0x3FFFFF )       ;
      break;
    case 18:
      n        = (instr & 0x7FFFFFF)       ;
      break;
    default:
      reg1     = (instr & 0x7C00000)  >> 22;
      imm      = (instr & 0x200000 )  >> 21;
      o        = (instr & 0x1FFFE0 )  >>  5;
      reg2     = (instr & 0X1F     )       ;
  }

}

/* the VM runs until this flag becomes 0 */
int running = 1;

/* evaluate the last decoded instruction */
void eval()
{
  switch( instrNum )
  {
    case 0:
      /* halt */
      printf( "stop\n" );
      running = 0;
      break;
    case 1:
      /* add   */
      if (imm == 1) {
        printf( "add r%d #%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] + o;
        break;
      }
      else {
        printf( "add r%d r%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] + regs[o];
        break;
      }
    case 2:
      /* sub   */
      if (imm == 1) {
        printf( "sub r%d #%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] - o;
        break;
      }
      else {
        printf( "sub r%d r%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] - regs[ o ];
        break;
      }
    case 3:
      /* mult  */
      if (imm == 1) {
        printf( "mult r%d #%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] * o;
        break;
      }
      else {
        printf( "mult r%d r%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] * regs[ o ];
        break;
      }
    case 4:
      /* div   */
      if (imm == 1) {
        printf( "div r%d #%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] / o;
        break;
      }
      else {
        printf( "div r%d r%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] / regs[ o ];
        break;
      }
    case 5:
      /* and   */
      if (imm == 1) {
        printf( "and r%d #%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] & o;
        break;
      }
      else {
        printf( "and r%d r%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] & regs[ o ];
        break;
      }
    case 6:
      /* or    */
      if (imm == 1) {
        printf( "or r%d #%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] | o;
        break;
      }
      else {
        printf( "or r%d r%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] | regs[ o ];
        break;
      }
    case 7:
      /* xor   */
      if (imm == 1) {
        printf( "xor r%d #%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] ^ o;
        break;
      }
      else {

      }
    case 8:
      /* shl   */
      if (imm == 1) {
        printf( "shl r%d #%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] << o;
        break;
      }
      else {
        printf( "shl r%d r%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] << regs[ o ];
        break;
      }
    case 9:
      /* shr   */
      if (imm == 1) {
        printf( "shr r%d #%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] >> o;
        break;
      }
      else {
        printf( "shr r%d r%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] >> regs[ o ];
        break;
      }
    case 10:
      /* slt   */
      if (imm == 1) {
        printf( "slt r%d #%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] < o;
        break;
      }
      else {
        printf( "slt r%d r%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] < regs[ o ];
        break;
      }
    case 11:
      /* sle   */
      if (imm == 1) {
        printf( "sle r%d #%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] <= o;
        break;
      }
      else {
        printf( "sle r%d r%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] <= regs[ o ];
        break;
      }
    case 12:
      /* seq   */
      if (imm == 1) {
        printf( "seq r%d #%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] == o;
        break;
      }
      else {
        printf( "seq r%d r%d r%d\n", reg1, o, reg2 );
        regs[ reg2 ] = regs[ reg1 ] == regs[ o ];
        break;
      }
    case 13:
      /* load  */
      printf( "load r%d #%d r%d\n", reg1, o, reg2);
      regs[ reg2 ] = regs[reg1 + o];
      break;

    case 14:
      /* store */
      printf( "store r%d #%d r%d\n", reg1, o, reg2);
      regs[ reg1 + o ] = regs[reg2];
      break;

    case 15:
      /* jmp   */
      printf( "jmp r%d r%d", reg1, imm );
      if (pc + imm > INSTR_LIM || program[pc+imm] == 0) {
        perror("Damn ! You jump too far !");
        break;
      }
      int instr = program[pc + imm];
      decode(instr);
      eval();
      break;
    case 16:
      /* braz  */
      printf("braz r%d r%d", reg1, imm);
      if (regs[reg1] == 0) {
        int instr = program[imm];
        decode(instr);
        eval();
      }
      break;
    case 17:
      /* branz */
      printf("branz r%d r%d", reg1, imm);
      if (regs[reg1] != 0) {
        int instr = program[imm];
        decode(instr);
        eval();
      }
      break;
    case 18:
      /* scall */
      printf("scall n%d", n);
      n = 0;
      switch(imm)
      {
        case 0:
          printf("Donnez un nombre : ");
          read(0,data,128);
          printf("\n");
          regs[reg1] = atoi(data);
          break;
        case 1:
          printf("data : %d", reg1);
          break;
      }
      break;
  }
}

/* display all registers as 4-digit hexadecimal words */
void showRegs()
{
  printf( "regs = \n" );
  int cpt=0;
  for(int i=0; i<NUM_REGS; i++ ) {
    printf( "%08X ", regs[ i ] );
    cpt++;
    if(cpt == 4)
    {
      printf(" \n");
      cpt = 0;
    }
  }
  printf( "\n" );
}

void run()
{
    while( running )
    {
      showRegs();
      unsigned int instr = fetch();
      decode( instr );
      eval();
    }
    showRegs();
}

int main( int argc, const char * argv[] )  {

  FILE* fichier = fopen(argv[1], "r");
  //printf("avant boucle de lecture\n");
  if(fichier == NULL){
    printf("Erreur lors de la lecture du fichier de programme \n");
  }

  else {
      //L'ouverture du fichier a reussi, on va pouvoir proceder a la traduction
      int i=0;
      char *chaine= (char *)malloc(TAILLE_MAX * sizeof(char));
      while (fgets(chaine, TAILLE_MAX, fichier) != NULL) {
        printf("%s", chaine); //afficher les lignes de la fiche d'instructio
        //program[i] = strtoul(program[i], chaine);
        program[i] = strtoul(chaine, NULL, 16);
        i++;
    }

  fclose(fichier);

  run();
  return 0;
  }
}
